File: backend/app/schemas/pico.py
from enum import StrEnum
from typing import Annotated

from pydantic import BaseModel, Field


class PicoConnectionStatus(StrEnum):
    DISCONNECTED = "DISCONNECTED"
    MOCK_CONNECTED = "MOCK_CONNECTED"
    CONNECTED = "CONNECTED"


class EstopState(StrEnum):
    UNKNOWN = "UNKNOWN"
    RELEASED = "RELEASED"
    ACTIVE = "ACTIVE"


class PinFunction(StrEnum):
    PAN_STEP = "PAN_STEP"
    PAN_DIR = "PAN_DIR"
    TILT_STEP = "TILT_STEP"
    TILT_DIR = "TILT_DIR"
    TRIGGER_SERVO_PWM = "TRIGGER_SERVO_PWM"
    ESTOP_IN = "ESTOP_IN"
    LIMIT_LEFT = "LIMIT_LEFT"
    LIMIT_RIGHT = "LIMIT_RIGHT"
    LIMIT_UP = "LIMIT_UP"
    LIMIT_DOWN = "LIMIT_DOWN"
    DRIVER_ENABLE = "DRIVER_ENABLE"
    UART_TX = "UART_TX"
    UART_RX = "UART_RX"
    UNUSED = "UNUSED"


class PinDirection(StrEnum):
    IN = "IN"
    OUT = "OUT"
    BIDIRECTIONAL = "BIDIRECTIONAL"
    UNUSED = "UNUSED"


class PinMode(StrEnum):
    GPIO = "GPIO"
    PWM = "PWM"
    UART = "UART"
    UNUSED = "UNUSED"


class PinValidationLevel(StrEnum):
    INFO = "INFO"
    WARNING = "WARNING"
    ERROR = "ERROR"
    CRITICAL = "CRITICAL"


class PicoTelemetry(BaseModel):
    connection_status: PicoConnectionStatus = PicoConnectionStatus.DISCONNECTED
    port: str | None = None
    baudrate: Annotated[int, Field(gt=0)] = 115200
    heartbeat_age_ms: Annotated[int, Field(ge=0)] | None = None
    firmware_version: str = "mock-pico-0.1"
    estop_state: EstopState = EstopState.UNKNOWN
    driver_enabled: bool = False
    pan_position_steps: int = 0
    tilt_position_steps: int = 0
    pan_limit_left: bool = False
    pan_limit_right: bool = False
    tilt_limit_up: bool = False
    tilt_limit_down: bool = False
    last_error: str | None = "mock_pico_not_connected"
    updated_at: float


class PicoStatus(BaseModel):
    mock_mode: bool = True
    telemetry: PicoTelemetry
    reason: str
    blocking_reasons: list[str]


class PicoPort(BaseModel):
    device: str
    label: str
    mock: bool = False


class PicoConnectRequest(BaseModel):
    port: str
    baudrate: Annotated[int, Field(gt=0)] = 115200


class PicoConnectionEvent(BaseModel):
    connection_status: PicoConnectionStatus
    port: str | None
    baudrate: int
    reason: str


class PicoDiscoveryPort(BaseModel):
    port: str
    description: str = "not_available"
    hwid: str | None = None
    vid: str | None = None
    pid: str | None = None
    serial_number: str | None = None
    manufacturer: str | None = None
    is_candidate: bool = False
    physical_command_enabled: bool = False
    no_physical_command_generated: bool = True


class PicoDiscoveryPortsResponse(BaseModel):
    ports: list[PicoDiscoveryPort]
    candidates_count: int = 0
    physical_command_enabled: bool = False
    no_physical_command_generated: bool = True


class PicoReadOnlyConnectRequest(BaseModel):
    port: str
    baudrate: Annotated[int, Field(gt=0)] = 115200
    read_only: bool = True


class PicoReadOnlyStatus(BaseModel):
    connected: bool = False
    selected_port: str | None = None
    baudrate: int = 115200
    rx_only: bool = True
    tx_disabled: bool = True
    serial_write_enabled: bool = False
    command_tx_enabled: bool = False
    last_seen_at: float | None = None
    heartbeat_seen: bool = False
    firmware_version: str | None = None
    telemetry_frames: int = 0
    parse_errors: int = 0
    dtr_rts_reset_risk: str = "reported_not_used_for_commands"
    warnings: list[str] = Field(default_factory=list)
    physical_command_enabled: bool = False
    no_physical_command_generated: bool = True


class PicoReadOnlyTelemetry(BaseModel):
    raw_line_sample: str | None = None
    parsed: dict = Field(default_factory=dict)
    heartbeat: bool = False
    firmware_version: str | None = None
    estop_state: str | None = None
    limit_states: dict = Field(default_factory=dict)
    motor_driver_state: dict = Field(default_factory=dict)
    warning_fault_state: dict = Field(default_factory=dict)
    no_command_generated: bool = True
    serial_write_enabled: bool = False
    command_tx_enabled: bool = False
    physical_command_enabled: bool = False
    no_physical_command_generated: bool = True


class PicoReadOnlyEvidence(BaseModel):
    evidence_id: str
    status: str
    created_at: float
    status_snapshot: PicoReadOnlyStatus
    latest_telemetry: PicoReadOnlyTelemetry
    port_inventory: list[PicoDiscoveryPort] = Field(default_factory=list)
    advisory_only: bool = True
    serial_write_enabled: bool = False
    command_tx_enabled: bool = False
    physical_command_enabled: bool = False
    no_physical_command_generated: bool = True


class PicoPermissionDiagnosis(BaseModel):
    port: str | None = None
    status: str = "not_available"
    blocker_class: str = "device_missing"
    user: str = "unknown"
    groups: list[str] = Field(default_factory=list)
    user_in_dialout: bool = False
    device_exists: bool = False
    device_mode: str | None = None
    device_owner: str | None = None
    device_group: str | None = None
    id_output: str = ""
    groups_output: str = ""
    ls_output: str = ""
    udevadm_output: str = ""
    dmesg_output: str = ""
    manual_recommendations: list[str] = Field(default_factory=list)
    serial_write_enabled: bool = False
    command_tx_enabled: bool = False
    physical_command_enabled: bool = False
    no_physical_command_generated: bool = True


class PicoProtocolLimitState(BaseModel):
    pan_left: bool = False
    pan_right: bool = False
    tilt_up: bool = False
    tilt_down: bool = False


class PicoProtocolFaultState(BaseModel):
    active: bool = False
    code: str | None = None
    message: str | None = None


class PicoProtocolTelemetry(BaseModel):
    protocol_name: str = "ISTIKLAL Serial Packet Protocol"
    protocol_version: int = 1
    pico_connected: bool = False
    telemetry_fresh: bool = False
    telemetry_missing: bool = True
    port: str | None = None
    last_heartbeat_age_ms: int | None = None
    last_packet_type: str | None = None
    last_packet_seq_id: int | None = None
    pan_deg: float | None = None
    tilt_deg: float | None = None
    x_steps: int | None = None
    y_steps: int | None = None
    driver_enabled: bool = False
    limit_state: PicoProtocolLimitState = Field(default_factory=PicoProtocolLimitState)
    fault_state: PicoProtocolFaultState = Field(default_factory=PicoProtocolFaultState)
    pose_source: str = "tracker_estimate"
    packet_parse_status: str = "no_packet"
    crc_status: str = "not_checked"
    physical_tx_disabled: bool = True
    serial_tx_enabled: bool = False
    physical_command_enabled: bool = False
    no_physical_command_generated: bool = True
    updated_at: float | None = None


class PicoProtocolPort(BaseModel):
    port: str
    description: str = "not_available"
    hwid: str | None = None
    is_candidate: bool = False
    no_physical_command_generated: bool = True


class PicoProtocolStatus(BaseModel):
    protocol_name: str = "ISTIKLAL Serial Packet Protocol"
    protocol_version: int = 1
    selected_port: str | None = None
    baudrate: int = 115200
    pico_connected: bool = False
    telemetry_fresh: bool = False
    telemetry_missing: bool = True
    latest_telemetry: PicoProtocolTelemetry = Field(default_factory=PicoProtocolTelemetry)
    discovered_ports: list[PicoProtocolPort] = Field(default_factory=list)
    packet_parse_status: str = "no_packet"
    crc_status: str = "not_checked"
    parse_errors: int = 0
    crc_failures: int = 0
    serial_tx_enabled: bool = False
    physical_tx_disabled: bool = True
    physical_command_enabled: bool = False
    no_physical_command_generated: bool = True


class PicoProtocolReadSampleRequest(BaseModel):
    sample_hex: str | None = None
    sample_base64: str | None = None
    sample_text: str | None = None


class PicoProtocolReadSampleResult(BaseModel):
    accepted: bool = True
    packets_parsed: int = 0
    errors: list[str] = Field(default_factory=list)
    remainder_len: int = 0
    latest_telemetry: PicoProtocolTelemetry = Field(default_factory=PicoProtocolTelemetry)
    physical_tx_disabled: bool = True
    physical_command_enabled: bool = False
    no_physical_command_generated: bool = True


class PinAssignment(BaseModel):
    pin_name: str
    physical_pin: Annotated[int, Field(gt=0)]
    function: PinFunction = PinFunction.UNUSED
    direction: PinDirection = PinDirection.UNUSED
    mode: PinMode = PinMode.UNUSED
    pwm_capable: bool = False
    uart_capable: bool = False
    note: str | None = None


class PinProfile(BaseModel):
    profile_name: str
    note: str
    final_approved: bool = False
    pins: list[PinAssignment]


class PinValidationIssue(BaseModel):
    level: PinValidationLevel
    code: str
    message: str
    pin_name: str | None = None
    function: PinFunction | None = None


class PinValidationResult(BaseModel):
    valid: bool
    can_apply: bool
    system_mode: str
    system_armed: bool
    issues: list[PinValidationIssue] = Field(default_factory=list)

File: frontend/src/types/pico.ts
export type PicoConnectionStatus = 'DISCONNECTED' | 'MOCK_CONNECTED' | 'CONNECTED'
export type EstopState = 'UNKNOWN' | 'RELEASED' | 'ACTIVE'
export type PinFunction =
  | 'PAN_STEP'
  | 'PAN_DIR'
  | 'TILT_STEP'
  | 'TILT_DIR'
  | 'TRIGGER_SERVO_PWM'
  | 'ESTOP_IN'
  | 'LIMIT_LEFT'
  | 'LIMIT_RIGHT'
  | 'LIMIT_UP'
  | 'LIMIT_DOWN'
  | 'DRIVER_ENABLE'
  | 'UART_TX'
  | 'UART_RX'
  | 'UNUSED'

export type PinDirection = 'IN' | 'OUT' | 'BIDIRECTIONAL' | 'UNUSED'
export type PinMode = 'GPIO' | 'PWM' | 'UART' | 'UNUSED'
export type PinValidationLevel = 'INFO' | 'WARNING' | 'ERROR' | 'CRITICAL'

export interface PicoTelemetry {
  connection_status: PicoConnectionStatus
  port: string | null
  baudrate: number
  heartbeat_age_ms: number | null
  firmware_version: string
  estop_state: EstopState
  driver_enabled: boolean
  pan_position_steps: number
  tilt_position_steps: number
  pan_limit_left: boolean
  pan_limit_right: boolean
  tilt_limit_up: boolean
  tilt_limit_down: boolean
  last_error: string | null
  updated_at: number
}

export interface PicoStatus {
  mock_mode: boolean
  telemetry: PicoTelemetry
  reason: string
  blocking_reasons: string[]
}

export interface PicoPort {
  device: string
  label: string
  mock: boolean
}

export interface PicoConnectionEvent {
  connection_status: PicoConnectionStatus
  port: string | null
  baudrate: number
  reason: string
}

export interface PicoDiscoveryPort {
  port: string
  description: string
  hwid: string | null
  vid: string | null
  pid: string | null
  serial_number: string | null
  manufacturer: string | null
  is_candidate: boolean
  physical_command_enabled: boolean
  no_physical_command_generated: boolean
}

export interface PicoDiscoveryPortsResponse {
  ports: PicoDiscoveryPort[]
  candidates_count: number
  physical_command_enabled: boolean
  no_physical_command_generated: boolean
}

export interface PicoReadOnlyStatus {
  connected: boolean
  selected_port: string | null
  baudrate: number
  rx_only: boolean
  tx_disabled: boolean
  serial_write_enabled: boolean
  command_tx_enabled: boolean
  last_seen_at: number | null
  heartbeat_seen: boolean
  firmware_version: string | null
  telemetry_frames: number
  parse_errors: number
  dtr_rts_reset_risk: string
  warnings: string[]
  physical_command_enabled: boolean
  no_physical_command_generated: boolean
}

export interface PicoReadOnlyTelemetry {
  raw_line_sample: string | null
  parsed: Record<string, unknown>
  heartbeat: boolean
  firmware_version: string | null
  estop_state: string | null
  limit_states: Record<string, unknown>
  motor_driver_state: Record<string, unknown>
  warning_fault_state: Record<string, unknown>
  no_command_generated: boolean
  serial_write_enabled: boolean
  command_tx_enabled: boolean
  physical_command_enabled: boolean
  no_physical_command_generated: boolean
}

export interface PicoReadOnlyEvidence {
  evidence_id: string
  status: string
  created_at: number
  status_snapshot: PicoReadOnlyStatus
  latest_telemetry: PicoReadOnlyTelemetry
  port_inventory: PicoDiscoveryPort[]
  advisory_only: boolean
  serial_write_enabled: boolean
  command_tx_enabled: boolean
  physical_command_enabled: boolean
  no_physical_command_generated: boolean
}

export interface PicoPermissionDiagnosis {
  port: string | null
  status: string
  blocker_class: string
  user: string
  groups: string[]
  user_in_dialout: boolean
  device_exists: boolean
  device_mode: string | null
  device_owner: string | null
  device_group: string | null
  id_output: string
  groups_output: string
  ls_output: string
  udevadm_output: string
  dmesg_output: string
  manual_recommendations: string[]
  serial_write_enabled: boolean
  command_tx_enabled: boolean
  physical_command_enabled: boolean
  no_physical_command_generated: boolean
}

export interface PicoProtocolLimitState {
  pan_left: boolean
  pan_right: boolean
  tilt_up: boolean
  tilt_down: boolean
}

export interface PicoProtocolFaultState {
  active: boolean
  code: string | null
  message: string | null
}

export interface PicoProtocolTelemetry {
  protocol_name: string
  protocol_version: number
  pico_connected: boolean
  telemetry_fresh: boolean
  telemetry_missing: boolean
  port: string | null
  last_heartbeat_age_ms: number | null
  last_packet_type: string | null
  last_packet_seq_id: number | null
  pan_deg: number | null
  tilt_deg: number | null
  x_steps: number | null
  y_steps: number | null
  driver_enabled: boolean
  limit_state: PicoProtocolLimitState
  fault_state: PicoProtocolFaultState
  pose_source: 'telemetry' | 'tracker_estimate' | 'fixture' | string
  packet_parse_status: string
  crc_status: string
  physical_tx_disabled: boolean
  serial_tx_enabled: boolean
  physical_command_enabled: boolean
  no_physical_command_generated: boolean
  updated_at: number | null
}

export interface PicoProtocolPort {
  port: string
  description: string
  hwid: string | null
  is_candidate: boolean
  no_physical_command_generated: boolean
}

export interface PicoProtocolStatus {
  protocol_name: string
  protocol_version: number
  selected_port: string | null
  baudrate: number
  pico_connected: boolean
  telemetry_fresh: boolean
  telemetry_missing: boolean
  latest_telemetry: PicoProtocolTelemetry
  discovered_ports: PicoProtocolPort[]
  packet_parse_status: string
  crc_status: string
  parse_errors: number
  crc_failures: number
  serial_tx_enabled: boolean
  physical_tx_disabled: boolean
  physical_command_enabled: boolean
  no_physical_command_generated: boolean
}

export interface PinAssignment {
  pin_name: string
  physical_pin: number
  function: PinFunction
  direction: PinDirection
  mode: PinMode
  pwm_capable: boolean
  uart_capable: boolean
  note: string | null
}

export interface PinProfile {
  profile_name: string
  note: string
  final_approved: boolean
  pins: PinAssignment[]
}

export interface PinValidationIssue {
  level: PinValidationLevel
  code: string
  message: string
  pin_name: string | null
  function: PinFunction | null
}

export interface PinValidationResult {
  valid: boolean
  can_apply: boolean
  system_mode: string
  system_armed: boolean
  issues: PinValidationIssue[]
}

export const PIN_FUNCTIONS: PinFunction[] = [
  'UNUSED',
  'PAN_STEP',
  'PAN_DIR',
  'TILT_STEP',
  'TILT_DIR',
  'TRIGGER_SERVO_PWM',
  'ESTOP_IN',
  'LIMIT_LEFT',
  'LIMIT_RIGHT',
  'LIMIT_UP',
  'LIMIT_DOWN',
  'DRIVER_ENABLE',
  'UART_TX',
  'UART_RX',
]
